In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [3]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

# Read the CSV and Perform Basic Data Cleaning

In [4]:
# https://help.lendingclub.com/hc/en-us/articles/215488038-What-do-the-different-Note-statuses-mean-

columns = [
    "loan_amnt", "int_rate", "installment", "home_ownership",
    "annual_inc", "verification_status", "issue_d", "loan_status",
    "pymnt_plan", "dti", "delinq_2yrs", "inq_last_6mths",
    "open_acc", "pub_rec", "revol_bal", "total_acc",
    "initial_list_status", "out_prncp", "out_prncp_inv", "total_pymnt",
    "total_pymnt_inv", "total_rec_prncp", "total_rec_int", "total_rec_late_fee",
    "recoveries", "collection_recovery_fee", "last_pymnt_amnt", "next_pymnt_d",
    "collections_12_mths_ex_med", "policy_code", "application_type", "acc_now_delinq",
    "tot_coll_amt", "tot_cur_bal", "open_acc_6m", "open_act_il",
    "open_il_12m", "open_il_24m", "mths_since_rcnt_il", "total_bal_il",
    "il_util", "open_rv_12m", "open_rv_24m", "max_bal_bc",
    "all_util", "total_rev_hi_lim", "inq_fi", "total_cu_tl",
    "inq_last_12m", "acc_open_past_24mths", "avg_cur_bal", "bc_open_to_buy",
    "bc_util", "chargeoff_within_12_mths", "delinq_amnt", "mo_sin_old_il_acct",
    "mo_sin_old_rev_tl_op", "mo_sin_rcnt_rev_tl_op", "mo_sin_rcnt_tl", "mort_acc",
    "mths_since_recent_bc", "mths_since_recent_inq", "num_accts_ever_120_pd", "num_actv_bc_tl",
    "num_actv_rev_tl", "num_bc_sats", "num_bc_tl", "num_il_tl",
    "num_op_rev_tl", "num_rev_accts", "num_rev_tl_bal_gt_0",
    "num_sats", "num_tl_120dpd_2m", "num_tl_30dpd", "num_tl_90g_dpd_24m",
    "num_tl_op_past_12m", "pct_tl_nvr_dlq", "percent_bc_gt_75", "pub_rec_bankruptcies",
    "tax_liens", "tot_hi_cred_lim", "total_bal_ex_mort", "total_bc_limit",
    "total_il_high_credit_limit", "hardship_flag", "debt_settlement_flag"
]

target = ["loan_status"]

In [5]:
# Load the data
file_path = Path('Resources/LoanStats_2019Q1.csv.zip')
df = pd.read_csv(file_path, skiprows=1)[:-2]
df = df.loc[:, columns].copy()

# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')

# Drop the null rows
df = df.dropna()

# Remove the `Issued` loan status
issued_mask = df['loan_status'] != 'Issued'
df = df.loc[issued_mask]

# convert interest rate to numerical
df['int_rate'] = df['int_rate'].str.replace('%', '')
df['int_rate'] = df['int_rate'].astype('float') / 100


# Convert the target column values to low_risk and high_risk based on their values
x = {'Current': 'low_risk'}   
df = df.replace(x)

x = dict.fromkeys(['Late (31-120 days)', 'Late (16-30 days)', 'Default', 'In Grace Period'], 'high_risk')    
df = df.replace(x)

df.reset_index(inplace=True, drop=True)

df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

# Split the Data into Training and Testing

In [6]:
# Create our features
X = df.copy()
X.drop(target, axis=1, inplace = True)


# Create our target
y = df[target]


In [7]:
#Binary encode columns
X = pd.get_dummies(X)

In [8]:
# Check the balance of our target values
y['loan_status'].value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [9]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test= train_test_split(X, 
                                                   y, 
                                                   random_state=1 
                                                   )

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [10]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [11]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [12]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Print the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier onely, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [13]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
brf = BalancedRandomForestClassifier(n_estimators = 100, random_state = 1)
brf.fit(X_train, y_train)
y_pred_brf = brf.predict(X_test)

In [14]:
# Calculated the balanced accuracy score
balanced_accuracy_score(y_test, y_pred_brf)

0.7885466545953005

In [15]:
# Display the confusion matrix
confusion_matrix(y_test, y_pred_brf)

array([[   71,    30],
       [ 2153, 14951]], dtype=int64)

In [16]:
# Print the imbalanced classification report

print(classification_report_imbalanced(y_test, y_pred_brf))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.03      0.70      0.87      0.06      0.78      0.60       101
   low_risk       1.00      0.87      0.70      0.93      0.78      0.62     17104

avg / total       0.99      0.87      0.70      0.93      0.78      0.62     17205



In [17]:
# List the features sorted in descending order by feature importance
importances = brf.feature_importances_
importances_df = pd.DataFrame(sorted(zip(importances, X.columns), reverse = True))
importances_df.set_index(importances_df[1], inplace = True)
importances_df.drop(columns =1, inplace = True)
importances_df.rename_axis('Feature Importance', inplace = True)
importances_df.rename(columns = {0: "Score"}, inplace = True)
importances_df


Score
Feature Importance                           
total_rec_prncp                      0.078768
total_pymnt                          0.058838
total_pymnt_inv                      0.056256
total_rec_int                        0.053555
last_pymnt_amnt                      0.050033
int_rate                             0.029670
issue_d_Jan-2019                     0.021129
installment                          0.019802
dti                                  0.017471
out_prncp_inv                        0.016858
total_bal_ex_mort                    0.016413
mths_since_recent_inq                0.015221
issue_d_Mar-2019                     0.015115
revol_bal                            0.014927
max_bal_bc                           0.014899
out_prncp                            0.014881
mo_sin_old_rev_tl_op                 0.014859
annual_inc                           0.014833
bc_util                              0.014614
tot_hi_cred_lim                      0.014488
mo_sin_old_il_acct                   0.013921
mths_since_rcnt_il                   0.013534
total_rev_hi_lim                     0.013365
avg_cur_bal                          0.013323
bc_open_to_buy                       0.013266
il_util                              0.013115
total_bc_limit                       0.013045
all_util                             0.013045
pct_tl_nvr_dlq                       0.013030
tot_cur_bal                          0.012856
...                                       ...
open_rv_12m                          0.004019
num_accts_ever_120_pd                0.003757
delinq_2yrs                          0.003083
home_ownership_RENT                  0.002913
home_ownership_OWN                   0.002825
verification_status_Not Verified     0.002205
application_type_Joint App           0.002091
pub_rec                              0.002015
verification_status_Verified         0.001933
application_type_Individual          0.001902
verification_status_Source Verified  0.001505
home_ownership_MORTGAGE              0.001459
pub_rec_bankruptcies                 0.001373
initial_list_status_w                0.001152
initial_list_status_f                0.001101
num_tl_90g_dpd_24m                   0.000996
collections_12_mths_ex_med           0.000715
home_ownership_ANY                   0.000496
chargeoff_within_12_mths             0.000273
tax_liens                            0.000000
recoveries                           0.000000
pymnt_plan_n                         0.000000
policy_code                          0.000000
num_tl_30dpd                         0.000000
num_tl_120dpd_2m                     0.000000
hardship_flag_N                      0.000000
delinq_amnt                          0.000000
debt_settlement_flag_N               0.000000
collection_recovery_fee              0.000000
acc_now_delinq                       0.000000

[95 rows x 1 columns]

### Easy Ensemble Classifier

In [18]:
# Train the EasyEnsembleClassifier
from imblearn.ensemble import EasyEnsembleClassifier

eec = EasyEnsembleClassifier(n_estimators = 100, random_state =1)
eec.fit(X_train, y_train)
y_pred_eec = eec.predict(X_test)

In [19]:
# Calculated the balanced accuracy score
balanced_accuracy_score(y_test, y_pred_eec)

0.9316600714093861

In [20]:
# Display the confusion matrix
print(confusion_matrix(y_test, y_pred_eec))

[[   93     8]
 [  983 16121]]


In [21]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred_eec))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.09      0.92      0.94      0.16      0.93      0.87       101
   low_risk       1.00      0.94      0.92      0.97      0.93      0.87     17104

avg / total       0.99      0.94      0.92      0.97      0.93      0.87     17205



### Which model had the best balanced accuracy score?

Easy Ensemble has the the best balanced accuracy score by far.  In fact much better than Random forest (.78) and the best of all the models (including the resampling techniques).

### Which model had the best recall score?

Random forest a .87 average, but .70 for high risk/.87 for low risk.  Easy Ensemble had a very high .94 average, and solid .92/.94.  Again, the best of all models.

### Which model had the best geometric mean score?

Easy Ensemble has the best geometric score too of all models. 

### What are the top three features?
Top 3 features are:  
total_rec_prncp 	0.078768
total_pymnt 	0.058838
total_pymnt_inv 	0.056256